### PR TITLE
Add smooth flow GrMhd AnalyticSolution

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

Adds an analytic solution to the GRMHD equations to PointwiseFunctions/AnalyticSolutions/GrMhd

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticSolutions/CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 add_subdirectory(Burgers)
 add_subdirectory(EinsteinSolutions)
+add_subdirectory(GrMhdSolutions)
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/CMakeLists.txt]: I don't think adding the suffix "Solutions" is necessary here. I actually need to change that for `NewtonianEulerSolutions` :)

File: src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/ConstantFlow.hpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+ * \f{align*}
+ * \rho(\vec{x},t) &= \rho, \\
+ * v_x(\vec{x},t) &= v_x,\\
+ * v_y(\vec{x},t) &= v_y,\\
+ * v_z(\vec{x},t) &= v_z,\\
+ * \epsilon(\vec{x}) &= \epsilon,\\
+ * \B_x(\vec{x},t) &= 0,\\
+ * \B_y(\vec{x},t) &= 0,\\
+ * \B_z(\vec{x},t) &= 0
+ * \f}
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/ConstantFlow.hpp]: The pressure should also be specified (for general EoSs, two thermodynamic variables are needed in addition to the rest-mass density.)

File: src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/ConstantFlow.hpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+  };
+
+  /// The specific internal energy of the fluid.
+  struct SpecificInternalEnergy {
+    using type = double;
+    static constexpr OptionString help = {
+        "The specific internal energy of the fluid."};
+  };
+
+  /// The lapse.
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/ConstantFlow.hpp]: Perhaps it should be documented that these 3+1 variables are not a solution to the Einstein equation using this flow as a source. These are constant numbers that make no impact on the right-hand side of the GRMHD equations, but we set them equal to something different from zero just for testing purposes.

File: src/PointwiseFunctions/AnalyticSolutions/CMakeLists.txt (hunk 1 above)
Comment (kidder) [reply src/PointwiseFunctions/AnalyticSolutions/CMakeLists.txt]: agreed, it is redundant

File: src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    static constexpr OptionString help = {
+        "The adiabatic exponent for the polytropic fluid."};
+  };
+
+  /// The amplitude of the rest mass density of the fluid.
+  struct DensityAmplitude {
+    using type = double;
+    static constexpr OptionString help = {
+        "The amplitude of the rest mass density of the fluid."};
+    static type upper_bound() { return 1.0; }
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp]: should have lower bound

File: src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    using type = std::array<double, 3>;
+    static constexpr OptionString help = {"The wave vector of the profile."};
+  };
+
+  /// The constant pressure throughout the fluid.
+  struct Pressure {
+    using type = double;
+    static constexpr OptionString help = {
+        "The constant pressure throughout the fluid."};
+  };
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp]: lower bound of zero?

File: src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  /// The adiabatic exponent for the polytropic fluid.
+  struct AdiabaticExponent {
+    using type = double;
+    static constexpr OptionString help = {
+        "The adiabatic exponent for the polytropic fluid."};
+  };
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp]: bounds?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/CMakeLists.txt [hunk 7]
```diff
@@ -1,4 +1,4 @@
-# Distributed under the MIT License.
+#Distributed under the MIT License.
```
Comment (kidder) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/CMakeLists.txt]: ?

File: src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  AdiabaticExponent::type adiabatic_exponent() const noexcept {
+    return adiabatic_exponent_;
+  }
+  DensityAmplitude::type density_amplitude() const noexcept {
+    return density_amplitude_;
+  }
+  std::array<double, 3> magnetic_field() const noexcept {
+    return magnetic_field_;
+  }
+  double lapse() const noexcept { return lapse_; }
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp]: why are these needed?

File: src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+  MeanVelocity::type mean_velocity_ =
+      make_array<3>(std::numeric_limits<double>::signaling_NaN());
+  WaveVector::type wavevector_ =
+      make_array<3>(std::numeric_limits<double>::signaling_NaN());
+  Pressure::type pressure_ = std::numeric_limits<double>::signaling_NaN();
+  AdiabaticExponent::type adiabatic_exponent_ =
+      std::numeric_limits<double>::signaling_NaN();
+  DensityAmplitude::type density_amplitude_ =
+      std::numeric_limits<double>::signaling_NaN();
+  std::array<double, 3> magnetic_field_ = make_array<3>(0.0);
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhdSolutions/SmoothFlow.hpp]: and if kept they need to be constexpr?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    for (size_t i = 0; i < 3; ++i) {
+      local_result += gsl::at(wavevector_, i) *
+                      (x.get(i) - make_with_value<DataType>(
+                                      x, t * gsl::at(mean_velocity_, i)));
+    }
+    return local_result;
+  }
+  ();
+
+  const DataType rest_mass_density =
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: make non-const and move into the tuple

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+      Scalar<DataType>{rest_mass_density};
+
+  auto velocity = make_with_value<tnsr::I<DataType, 3>>(x, 0.0);
+  for (size_t i = 0; i < 3; ++i) {
+    velocity.get(i) = gsl::at(mean_velocity_, i);
+  }
+  get<hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>>(result) =
+      std::move(velocity);
+
+  get<hydro::Tags::SpecificInternalEnergy<DataType>>(result) =
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: Don't introduce `specific_internal_energy`, just compute in place after computing `rest_mass_density` but before moving `rest_mass_density`

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+      Scalar<DataType>(specific_internal_energy);
+
+  // Magnetic field is not set because it is identically zero,
+  // and `result` is initialized with all primitive variables equal to zero.
+
+  get<hydro::Tags::Pressure<DataType>>(result) =
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: `get(get<...>) = pressure_`

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  ();
+
+  const DataType rest_mass_density =
+      1.0 + density_amplitude_ * sin(k_dot_x_minus_vt);
+  const DataType specific_internal_energy =
+      pressure_ / (adiabatic_exponent_ - 1.0) / rest_mass_density;
+  get<hydro::Tags::RestMassDensity<DataType>>(result) =
+      Scalar<DataType>{rest_mass_density};
+
+  auto velocity = make_with_value<tnsr::I<DataType, 3>>(x, 0.0);
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: don't declare `velocity`, just loop over the components in the tuple
`get<hydro...>(result).get(i) = gsl::at(mean_velocity_, i)`

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/Test_SmoothFlow.cpp [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.AnalyticSolutions.GrMhd.SmoothFlow",
+                  "[Unit][PointwiseFunctions]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "PointwiseFunctions/AnalyticSolutions/GrMhd"};
+
+  test_create_from_options();
+  test_serialize();
+  test_move();
+
+  // test_variables(std::numeric_limits<double>::signaling_NaN());
```
Comment (kidder) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/Test_SmoothFlow.cpp]: ?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,16 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(LIBRARY GrMhd)
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/CMakeLists.txt]: The name of the library should be `GrMhdSolutions`

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+#include "Utilities/MakeWithValue.hpp"
+/// \cond
+namespace grmhd {
+namespace Solutions {
+
+SmoothFlow::SmoothFlow(MeanVelocity::type mean_velocity,
+                       WaveVector::type wavevector,
+                       const Pressure::type pressure,
+                       const AdiabaticExponent::type adiabatic_exponent,
+                       const DensityAmplitude::type density_amplitude)
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: constructor can be `noexcept`

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+      local_result += gsl::at(wavevector_, i) *
+                      (x.get(i) - make_with_value<DataType>(
+                                      x, t * gsl::at(mean_velocity_, i)));
+    }
+    return local_result;
+  }
+  ();
+  get(get<hydro::Tags::RestMassDensity<DataType>>(result)) =
+      1.0 + density_amplitude_ * sin(k_dot_x_minus_vt);
+  get<hydro::Tags::SpecificInternalEnergy<DataType>>(result) = Scalar<DataType>(
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: Can't you just assign to the datavector directly with another get call?

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+      pressure_ / (adiabatic_exponent_ - 1.0) /
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: `pressure_ / ((a - 1.0) * ...)`

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+  get(get<Tags::dt<hydro::Tags::RestMassDensity<DataType>>>(result)) =
+      -density_amplitude_ * k_dot_v * cos(k_dot_x_minus_vt);
+
+  get(get<Tags::dt<hydro::Tags::SpecificInternalEnergy<DataType>>>(result)) =
+      -pressure_ /
+      ((adiabatic_exponent_ - 1.0) *
+       square(1.0 + density_amplitude_ * sin(k_dot_x_minus_vt))) *
+      get(get<Tags::dt<hydro::Tags::RestMassDensity<DataType>>>(result));
+
+  // The time derivative of the spatial velocity is not set because it is
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: Could you just have one comment that lists `the spatial velocity, magnetic field, and pressure are not...`?

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+ * An analytic solution to the 3-D GrMhd system. The user specifies the mean
+ * flow velocity of the fluid, the wavevector of the density profile, and the
+ * amplitude \f$A\f$ of the density profile. The magnetic field is taken to be
+ * zero everywhere. In Cartesian coordinates \f$(x, y, z)\f$, and using
+ * dimensionless units, the primitive quantities at a given time \f$t\f$ are
+ * then
+ *
+ * \f{align*}
+ * \rho(\vec{x},t) &= 1 + A \sin(\vec{k}\cdot(\vec{x} - \vec{v}t)) \\
+ * v_x(\vec{x},t) &= v_x,\\
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: why not write this as `\vec{v}(\vec{x}, t) & = [0, 0, 0]^{T}` (the transpose)? Same with the B field

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+ * v_y(\vec{x},t) &= v_y,\\
+ * v_z(\vec{x},t) &= v_z,\\
+ * P(\vec{x},t) &= P, \\
+ * \epsilon(\vec{x}, t) &= \frac{P}{(\gamma - 1)\rho}\\
+ * B_x(\vec{x},t) &= 0,\\
+ * B_y(\vec{x},t) &= 0,\\
+ * B_z(\vec{x},t) &= 0
+ * \f}
+ *
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: decrement line count

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,15 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(LIBRARY "Test_GrMhd")
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/CMakeLists.txt]: GrMhdSolutions

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/Test_SmoothFlow.cpp [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+  }
+};
+
+void test_create_from_options() noexcept {
+  test_creation<grmhd::Solutions::SmoothFlow>(
+      "  MeanVelocity: [0.1, -0.2, 0.3]\n"
+      "  WaveVector: [-0.13, -0.54, 0.04]\n"
+      "  Pressure: 1.23\n"
+      "  AdiabaticExponent: 1.4\n"
+      "  DensityAmplitude: 0.75");
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/Test_SmoothFlow.cpp]: Can you actually test the values you got?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm fine with you squashing right away

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+  /// The adiabatic exponent for the polytropic fluid.
+  struct AdiabaticExponent {
+    using type = double;
+    static constexpr OptionString help = {
+        "The adiabatic exponent for the polytropic fluid."};
+    static type lower_bound() { return 1.0; }
+  };
+
+  /// The amplitude of the rest mass density of the fluid.
+  struct DensityAmplitude {
```
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: PerturbationAmplitude

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+    using type = double;
+    static constexpr OptionString help = {
+        "The amplitude of the rest mass density of the fluid."};
+    static type lower_bound() { return 0.0; }
```
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: can be made -1

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+  using variables_t =
+      tmpl::list<hydro::Tags::RestMassDensity<DataType>,
+                 hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>,
+                 hydro::Tags::SpecificInternalEnergy<DataType>,
+                 hydro::Tags::Pressure<DataType>,
+                 hydro::Tags::MagneticField<DataType, 3, Frame::Inertial>>;
+
+  template <typename DataType>
+  using dt_variables_t = tmpl::list<
+      Tags::dt<hydro::Tags::RestMassDensity<DataType>>,
```
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: use wrap_tags_in metafunction

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+SmoothFlow::variables(const tnsr::I<DataType, 3>& x, const double t,
+                      SmoothFlow::dt_variables_t<DataType> /*meta*/) const
+    noexcept {
+  // Explicitly set all variables to zero:
+  auto result = make_with_value<
+      tuples::tagged_tuple_from_typelist<SmoothFlow::dt_variables_t<DataType>>>(
+      x, 0.0);
+
+  // Compute the phase: k * (x - vt)
+  const DataType k_dot_x_minus_vt = [&x, &t, this ]() noexcept {
```
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: factor this out

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+      local_result += gsl::at(wavevector_, i) *
+                      (x.get(i) - make_with_value<DataType>(
+                                      x, t * gsl::at(mean_velocity_, i)));
+    }
+    return local_result;
+  }
+  ();
+
+  // Compute the angular frequency: k * v
+  const DataType k_dot_v = [&x, this ]() noexcept {
```
Comment (mar-celine) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: can be made double - check for dot product function (std::inner_product)

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/Test_SmoothFlow.cpp [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+      {{{-15., 15.}}},
+      std::make_tuple(mean_velocity, wave_vector, pressure, adiabatic_exponent,
+                      density_amplitude),
+      used_for_size);
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.AnalyticSolutions.GrMhd.SmoothFlow",
+                  "[Unit][PointwiseFunctions]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
```
Comment (mar-celine) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GrMhd/Test_SmoothFlow.cpp:120]: make issue for generically numerically testing solutions

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp [hunk 30]
```diff
@@ -88,36 +92,17 @@ SmoothFlow::variables(const tnsr::I<DataType, 3>& x, const double t,
...
-    auto local_result = make_with_value<DataType>(x, 0.0);
-    for (size_t i = 0; i < 3; ++i) {
-      local_result += gsl::at(wavevector_, i) *
-                      make_with_value<DataType>(x, gsl::at(mean_velocity_, i));
-    }
-    return local_result;
-  }
-  ();
-
+  const double k_dot_v = std::inner_product(
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: how about saving this as a const member variable?  then it could be used on line 48 to initialize `result` making the function simpler.

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
There's a IWYU issue (it's correct), other than that I'm happy with this. Once @kidder is you can squash :)

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp (hunk 30 above)
Comment (mar-celine) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: I think this would make the class not assignable - should I mark the copy assignment and move assignment operators as delete? Or should I make k_dot_v_ non const? wavevector_ and mean_velocity_ are currently not const

Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: Make it a member variable but don't make it const. We don't want const member variables, as you say.

Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: You could also write custom copy and move semantics, but that seems like more work than it's worth for this class :)

Comment (kidder) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp]: yeah, just make it non-const

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp [hunk 31]
```diff
@@ -151,14 +146,18 @@ class SmoothFlow {
       std::numeric_limits<double>::signaling_NaN();
   PerturbationAmplitude::type perturbation_amplitude_ =
       std::numeric_limits<double>::signaling_NaN();
+  // The angular frequency.
+  double k_dot_v_ = std::inner_product(
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: I don't see how this can work here...

Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: Initialization occurs in order as far as I know, and this would be called at the end of the list initialization. @wthrowe is my memory correct?

Comment (kidder) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: won't you get an FPE if it uses the default values above?

Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: I believe all of these statements are true, and in fact I observe an FPE if I run the test on my machine (when default-constructing in the option parser, it looks like).

Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp]: Ah, yep. The default constructor would blow up... :(

**Comment by kidder** [comment]
@gsb76  an option help string is too long

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]
